### PR TITLE
work with phantomjs

### PR DESCRIPTION
**Pull request by calvinmetcalf**
fix for crypto-browserify/crypto-browserify#143

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+      env: TEST_SUITE=test
+    - node_js: '0.12'
+      env: TEST_SUITE=test
+    - node_js: '5'
+      env: TEST_SUITE=test
+    - node_js: '4'
+      env: TEST_SUITE=test
+    - node_js: '4'
+      env: TEST_SUITE=phantom
+script: "npm run-script $TEST_SUITE"
```
Comment (dcousens) [inline .travis.yml]: how was this not on travis already? damn

File: browser.js [hunk 2]
```diff
@@ -14,11 +14,14 @@ if (crypto && crypto.getRandomValues) {
 
 function randomBytes (size, cb) {
   // in browserify, this is becomes an extended UInt8Array
-  var bytes = new Buffer(size)
+  var rawBytes = new global.UInt8Array(size)
 
   // This will not work in older browsers.
   // See https://developer.mozilla.org/en-US/docs/Web/API/window.crypto.getRandomValues
-  crypto.getRandomValues(bytes)
+  crypto.getRandomValues(rawBytes)
```
Comment (dcousens) [inline browser.js]: so, my understanding is `phantomjs` didn't like giving `crypto.getRandomValues` a `Buffer` (aka, extended `UInt8Array`).
But, it is OK with initializing a `Buffer` with a `UInt8Array`.

Right?

Comment (calvinmetcalf) [inline browser.js]: I think so but I havn't been actually able to test it successfully on phantomjs

Comment (calvinmetcalf) [inline browser.js]: yes so the deal is that not every version of buffer actually is based off a uint8array if it isn't supported by that browser

**Comment by calvinmetcalf** [comment]
yeah there was a typo in it was why it was failing

File: test.js [hunk 3]
```diff
@@ -33,13 +33,17 @@ test('async', function (t) {
 if (process.browser) {
   test('requesting to much throws', function (t) {
     t.plan(1)
-    t.throws(randomBytes.bind(null, 65537))
```
Comment (calvinmetcalf) [inline test.js]: phantomjs doesn't have Function.prototype.bind

File: browser.js (hunk 2 above)
Comment (dcousens) [inline browser.js]: @feross could you pipe in?

Comment (feross) [inline browser.js]: In browsers that support Typed Arrays *and* have a correct implementation, `Buffer` will be a subclass of `Uint8Array`. In older or broken browsers, `Buffer` will just be an `Object`. You can see the exactly support list here: https://github.com/feross/buffer/blob/master/index.js#L20-L49

It's my understanding that `crypto.getRandomValues` absolutely must be given a `Uint8Array`, so it's expected to fail in browsers without support. Does the version of phantomjs you're using have support?

You can try logging out the value of `Buffer.TYPED_ARRAY_SUPPORT` to see if Buffer is actually using the `Uint8Array` implementation.

File: test.js (hunk 3 above)
Comment (dcousens) [inline test.js]: Really?
Isn't that ES5?

Comment (calvinmetcalf) [inline test.js]: i know

File: browser.js (hunk 2 above)
Comment (calvinmetcalf) [inline browser.js]: so it needs to be an integer based typed array, anything that starts with Int or Uint and this dues fix it

**Comment by calvinmetcalf** [comment]
I'm going to add a travis badge and then sqash this

**Comment by calvinmetcalf** [comment]
rebased

**Comment by calvinmetcalf** [comment]
currently this does cause an extra copy, but feross/buffer#83 handles that

**Comment by calvinmetcalf** [comment]
ok so that pull didn't make sense on buffer so updated this to avoid the extra copy

File: browser.js [hunk 4]
```diff
@@ -13,12 +13,19 @@ if (crypto && crypto.getRandomValues) {
 }
 
 function randomBytes (size, cb) {
-  // in browserify, this is becomes an extended UInt8Array
-  var bytes = new Buffer(size)
+  // phantomjs needs to throw
+  if (size > 65536) {
+    throw new Error('requested too many random bytes')
+  }
```
Comment (dcousens) [inline browser.js]: nit: could be 1 line

File: browser.js [hunk 5]
```diff
@@ -13,12 +13,19 @@ if (crypto && crypto.getRandomValues) {
...
+  // in case browserify  isn't use the Uint8Array version
```
Comment (dcousens) [inline browser.js]: nit: `s/use/using`

File: browser.js [hunk 6]
```diff
@@ -13,12 +13,19 @@ if (crypto && crypto.getRandomValues) {
...
+  var rawBytes = new global.Uint8Array(size)
 
   // This will not work in older browsers.
   // See https://developer.mozilla.org/en-US/docs/Web/API/window.crypto.getRandomValues
-  crypto.getRandomValues(bytes)
+  crypto.getRandomValues(rawBytes)
+
+  // phantomjs doesn't like a buffer being passed here
+  var bytes = new Buffer(rawBytes.buffer)
```
Comment (dcousens) [inline browser.js]: does this mean we are now always making an extra copy?

**Comment by dcousens** [comment]
LGTM

File: browser.js (hunk 6 above)
Comment (calvinmetcalf) [inline browser.js]: no this avoids the copy

File: browser.js (hunk 4 above)
Comment (calvinmetcalf) [inline browser.js]: I thought we always used curly brases

**Comment by calvinmetcalf** [comment]
@dcousens fixed the nits

**Comment by dcousens** [comment]
Great work @calvinmetcalf :+1:

**Comment by calvinmetcalf** [comment]
published

**Comment by substack** [comment]
This patch broke randombytes in real, non-phantom browsers.

**Pull request merged**